File: lib/web/isMatchPathname.ts
export function isMatchPathname(pathname: string, matcher: string | string[]): boolean {
	if (!pathname || !matcher || matcher?.length < 1) {
		throw new Error('`url` and `matcher` must be set');
	}

	const matcherSet = typeof matcher === 'string' ? [matcher] : matcher;
	const realPathname = pathname.split('?')[0].replace(/\/$/, '');

	return matcherSet.some((m) => {
		if (m === '*' || m === '/*') {
			return true;
		}

		const realMatcher = m.replace(/\*$/, '');

		return m.endsWith('*') ? realPathname.startsWith(realMatcher) : realMatcher === realPathname;
	});
}

File: packages/python/qsu/web/isMatchPathname.py
import re
from typing import List, Union


def isMatchPathname(pathname: str, matcher: Union[str, List[str]]) -> bool:
	if not pathname or not matcher or (matcher is not None and len(matcher) < 1):
		raise ValueError('`url` and `matcher` must be set')

	matcherSet = [matcher] if isinstance(matcher, str) else matcher
	realPathname = re.sub(r'/$', '', pathname.split('?')[0])

	for m in matcherSet:
		if m == '*' or m == '/*':
			return True

		realMatcher = re.sub(r'\*$', '', m)

		if m.endswith('*'):
			if realPathname.startswith(realMatcher):
				return True
		elif realMatcher == realPathname:
			return True

	return False
